File: app/src/main/java/org/wikipedia/util/MathUtil.java
package org.wikipedia.util;

public final class MathUtil {

    private static final int PERCENTAGE_BASE = 100;

    public static float constrain(float f, float min, float max) {
        return Math.min(Math.max(min, f), max);
    }

    public static class Averaged<T extends Number> {
        private double sampleSum;
        private int sampleSize;

        public void addSample(T sample) {
            sampleSum += sample.doubleValue();
            ++sampleSize;
        }

        public double getAverage() {
            return sampleSize == 0 ? 0 : sampleSum / sampleSize;
        }

        public void reset() {
            sampleSum = 0;
            sampleSize = 0;
        }
    }

    public static float percentage(float numerator, float denominator) {
        return numerator / denominator * PERCENTAGE_BASE;
    }

    private MathUtil() {
    }
}

File: app/src/main/java/org/wikipedia/util/MathUtil.kt
package org.wikipedia.util

object MathUtil {
    private const val PERCENTAGE_BASE = 100

    fun percentage(numerator: Float, denominator: Float): Float {
        return numerator / denominator * PERCENTAGE_BASE
    }

    class Averaged<T : Number?> {
        private var sampleSum = 0.0
        private var sampleSize = 0
        fun addSample(sample: T) {
            sampleSum += sample!!.toDouble()
            ++sampleSize
        }

        val average: Double
            get() = if (sampleSize == 0) 0.0 else sampleSum / sampleSize

        fun reset() {
            sampleSum = 0.0
            sampleSize = 0
        }
    }
}
